libcaf_core/caf/delegate.test.cpp

```diff
@@ -0,0 +1,109 @@
+#include "caf/test/test.hpp"
+
+#include "caf/event_based_actor.hpp"
+#include "caf/scoped_actor.hpp"
+#include "caf/sec.hpp"
+#include "caf/send.hpp"
+
+namespace fixture = caf::test::fixture;
+
+WITH_FIXTURE(fixture::deterministic) {
+
+#ifdef CAF_ENABLE_EXCEPTIONS
+TEST("delegates pass on the response properly to the sender") {
+  using namespace caf;
+  auto count = std::make_shared<int32_t>(0);
+  auto worker = sys.spawn([count] {
+    return behavior{
+      [count](int32_t value) {
+        *count = *count + value;
+        return *count;
+      },
+    };
+  });
+  auto delegator = sys.spawn(
+    [](event_based_actor* self, caf::actor worker) {
+      return behavior{
+        [self, worker](int32_t x) {
+          auto promise = self->make_response_promise<int>();
+          return promise.delegate(worker, x);
+        },
+        [self, worker](std::string x) {
+          auto promise = self->make_response_promise<error>();
+          return promise.delegate(worker, x);
+        },
+      };
+    },
+    worker);
+  SECTION("handled messages are received by client") {
+    auto client = sys.spawn([delegator](event_based_actor* self) {
+      self->send(delegator, 2);
+      self->send(delegator, 3);
+      return behavior{[](int32_t) {}};
+    });
+    expect<int32_t>().with(2).from(client).to(delegator);
+    expect<int32_t>().with(2).from(client).to(worker);
+    expect<int32_t>().with(2).from(worker).to(client);
+    check_eq(*count, 2);
+    expect<int32_t>().with(3).from(client).to(delegator);
+    expect<int32_t>().with(3).from(client).to(worker);
+    expect<int32_t>().with(5).from(worker).to(client);
+    check_eq(*count, 5);
+  }
+  SECTION("unhandled messages respond with error") {
+    auto client = sys.spawn([count, delegator](event_based_actor* self) {
+      self->send(delegator, "2");
+      return behavior{[](int32_t) {}};
+    });
+    expect<std::string>().with("2").from(client).to(delegator);
+    expect<std::string>().with("2").from(client).to(worker);
+    expect<error>().with(sec::unexpected_message).from(worker).to(client);
+  }
+  SECTION("consecutive messages are handled after unhandled message") {
+    auto client = sys.spawn([count, delegator](event_based_actor* self) {
+      self->send(delegator, "2");
+      self->send(delegator, 3);
+      self->send(delegator, 4);
+      return behavior{[](int32_t) {}};
+    });
+    expect<std::string>().with("2").from(client).to(delegator);
+    expect<std::string>().with("2").from(client).to(worker);
+    expect<error>().with(sec::unexpected_message).from(worker).to(client);
+    expect<int32_t>().with(3).from(client).to(delegator);
+    expect<int32_t>().with(3).from(client).to(worker);
+    expect<int32_t>().with(3).from(worker).to(client);
+    check_eq(*count, 3);
+    expect<int32_t>().with(4).from(client).to(delegator);
+    expect<int32_t>().with(4).from(client).to(worker);
+    expect<int32_t>().with(4).from(worker).to(client);
+    check_eq(*count, 7);
+  }
```

Comment:
This test makes no sense to me. The `client` does not override the error handler. Hence, it will terminate after receiving the unexpected message error. Consequently, Line 80 and line 84 can never pass, since `client` will close its mailbox when shutting down and the response from `worker` will be discarded.

In general, this test doesn't seem to cover anything that we haven't covered before. I'd just remove it unless there's an edge case that you have tried to test here and that I'm missing.

Reply:
In the issue description it was mentioned that 

> If actor B1 does not have a handler for this message and it falls into the deadletter (in akka terminology), the original sender actor in the chain on side A (A1) hangs and is no longer able to receive any messages.

I attempted to reproduce this as from my understanding the behavior for the actor would be to receive messages. But, if we are expecting that no further message will be received by the `client` , the test can only cover one more scenario that has not been tested and can be added, that is to check that the error handler for the delegate has been called. Shall I rewrite the tests to reproduce that scenario instead?

Reply:
I think we have the scenario in the ticket covered with test in line 59, no? If that test is passing, I think the comment by tariksmajic is pointing to the right direction. In the setup described by the original poster, the actor might not actually be hanging but in fact shut down properly.

To double-check this, add another actor to the setup for line 59 that monitors the client. After the client receives the error, we want the actor (a good name might be `observer`) to receive the `down_msg`.